### ISSUE TITLE
Keras assinging weights from pretrained model?

### ISSUE THREAD

**Issue by akbari59** [keras-team/keras#8957]
I want to load the weights of a LSTM model from a pre-trained model. I wrote the following code.
```
json_file = open(PRETRAIN_MODEL_FILENAME , "r")
loaded_model_json = json_file.read()
json_file.close()
loaded_model = model_from_json(loaded_model_json)
loaded_model.load_weights(PRETRAIN_MODEL_FILENAME + "-weights.h5")
print("Loaded model from disk")

embedding_weights = loaded_model.layers[1].get_weights()
lstm_weights = loaded_model.layers[2].get_weights()
softmax_weights = loaded_model.layers[3].get_weights()

print('evaluating loaded model...')
loaded_model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])
loaded_model.summary()
score = loaded_model.evaluate(x_test, y_test, verbose=0)
print("%s: %.2f%%" % (loaded_model.metrics_names[1], score[1]*100))
```

Now I want to create a new model with similar architecture and weights so I did as follows
```
embedding_layer = Embedding(len(word_index) + 1,
                        EMBEDDING_DIM,
                        weights=embedding_weights,
                        input_length=MAX_SEQUENCE_LENGTH,
                        trainable=True)

sequence_input = Input(shape=(MAX_SEQUENCE_LENGTH,), dtype='int32')
embedded_sequences = embedding_layer(sequence_input)
l_lstm = Bidirectional(LSTM(100), weights=lstm_weights)(embedded_sequences)
preds = Dense(num_output_categories, activation='softmax', 
weights=softmax_weights)(l_lstm)
model = Model(sequence_input, preds)
#model.load_weights(PRETRAIN_MODEL_FILENAME + "-weights.h5", by_name=True)


model.compile(loss='categorical_crossentropy',
          optimizer='rmsprop',
          metrics=['acc'])
print("Evaluating new model with pretrained weights")
score = model.evaluate(x_test, y_test, verbose=1)
model.summary()
print("%s: %.2f%%" % (model.metrics_names[1], score[1]*100))
```

While I used the same model and same wights, I cannot get similar results as when I load the weights using Keras. Where I do wrong or keras does not support this way of loading weights?
The output of above code is as follows:
```
Loaded model from disk
evaluating loaded model...
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         (None, 1000)              0         
_________________________________________________________________
embedding_1 (Embedding)      (None, 1000, 100)         7034500   
_________________________________________________________________
bidirectional_1 (Bidirection (None, 200)               160800    
_________________________________________________________________
dense_1 (Dense)              (None, 4)                 804       
=================================================================
Total params: 7,196,104
Trainable params: 7,196,104
Non-trainable params: 0
_________________________________________________________________
acc: 90.37%
Evaluating new model with pretrained weights
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         (None, 1000)              0         
_________________________________________________________________
embedding_1 (Embedding)      (None, 1000, 100)         7034500   
_________________________________________________________________
bidirectional_1 (Bidirection (None, 200)               160800    
_________________________________________________________________
dense_1 (Dense)              (None, 4)                 804       
=================================================================
Total params: 7,196,104
Trainable params: 7,196,104
Non-trainable params: 0
_________________________________________________________________
acc: 24.51%

```

**Comment by tomastheod-ITI** [comment]
@akbari59 First compile the model and then load the weights. When you compile a model, the weights are reset. The 24.51% accuracy when you have 4 classes confirms that your model is making completely random predictions.

**Comment by akbari59** [comment]
@tomastheod-ITI Thanks. you mean that the input to layer function call does not affect, if I compile the model? (the weight in following function, for example)
`l_lstm = Bidirectional(LSTM(100), weights=lstm_weights)(embedded_sequences)
`

**Comment by tomastheod-ITI** [comment]
@akbari59 Sorry, I completely missed your point! Your method seems fine. However, the second approach produces completely random results. After some experiments with different layers, I found out that the Bidirectional layer does not take into account the given weights.
```
from keras.models import Model
from keras.layers import Input, Dense, LSTM, Bidirectional
import numpy as np


def model_1():

    model_in = Input(shape=(30, 10))

    x = Bidirectional(LSTM(50))(model_in)

    x = Dense(units=2, activation='softmax')(x)

    model_out = x


    model = Model(inputs=model_in, outputs=model_out)

    model.compile(loss='binary_crossentropy', optimizer='sgd')

    model.summary(line_length=100)


    return model


def model_2(mod):

    model_in = Input(shape=(30, 10))

    w1 = mod.layers[1].get_weights()
    x = Bidirectional(LSTM(50), weights=w1)(model_in)

    w2 = mod.layers[2].get_weights()
    x = Dense(units=2, activation='softmax', weights=w2)(x)

    model_out = x


    model = Model(inputs=model_in, outputs=model_out)

    model.compile(loss='binary_crossentropy', optimizer='sgd')

    model.summary(line_length=100)


    return model


model1 = model_1()
model2 = model_2(model1)

for i in range(len(model1.layers)):

    if len(model1.layers[i].get_weights()) > 0:

        w1 = model1.layers[i].get_weights()
        w2 = model2.layers[i].get_weights()

        for j in range(len(w1)):

            print '{0} == {1} | {2}'.\
            format(model1.layers[i].name, model2.layers[i].name, np.all(w1[j] == w2[j]))
```
We define model1 and use its weights to initialize model2. However, the Bidirectional layer doesn't have the same weights in both cases. The output is this:
```
bidirectional_1 == bidirectional_2 | False
bidirectional_1 == bidirectional_2 | False
bidirectional_1 == bidirectional_2 | True
bidirectional_1 == bidirectional_2 | False
bidirectional_1 == bidirectional_2 | False
bidirectional_1 == bidirectional_2 | True
dense_1 == dense_2 | True
dense_1 == dense_2 | True
```
Only the biases in the Bidirectional layer are the same.

**Comment by akbari59** [comment]
@tomastheod-ITI Thanks for your explanation; for now I am using load_weights method with by_name input. But would be good to have the second option too.

**Comment by gtesei** [comment]
That's interesting. Looks like it happens only with  the **Bidirectional** wrapper. 
For example,  here below same code with LSTM only. 

```python
from keras.models import Model
from keras.layers import Input, Dense, LSTM, Bidirectional
import numpy as np

def model_1():
    model_in = Input(shape=(30, 10))
    x = LSTM(50)(model_in)
    x = Dense(units=2, activation='softmax')(x)
    model_out = x
    model = Model(inputs=model_in, outputs=model_out)
    model.compile(loss='binary_crossentropy', optimizer='sgd')
    model.summary(line_length=100)
    return model

def check_model(model1,model2):
    for i in range(len(model1.layers)):
        if len(model1.layers[i].get_weights()) > 0:
            w1 = model1.layers[i].get_weights()
            w2 = model2.layers[i].get_weights()
            for j in range(len(w1)):
                print('{0} == {1} | {2}'.format(model1.layers[i].name, 
                      model2.layers[i].name, 
                      np.all(w1[j] == w2[j])))


def model_2(mod):
    model_in = Input(shape=(30, 10))
    w1 = mod.layers[1].get_weights()
    x = LSTM(50,weights=w1)(model_in)
    w2 = mod.layers[2].get_weights()
    x = Dense(units=2, activation='softmax', weights=w2)(x)
    model_out = x
    model = Model(inputs=model_in, outputs=model_out)
    model.compile(loss='binary_crossentropy', optimizer='sgd')
    model.summary(line_length=100)
    return model

model1 = model_1()
#check_model(model1,model1)
model2 = model_2(model1)
#check_model(model2,model2)
check_model(model1,model2)
```

Output:
```
lstm_22 == lstm_23 | True
lstm_22 == lstm_23 | True
lstm_22 == lstm_23 | True
dense_28 == dense_29 | True
dense_28 == dense_29 | True
```

**Comment by dnajera27** [comment]
I am encountering this same issue. Is there a known workaround?
